### ISSUE TITLE
Page-load performance optimisation

### ISSUE THREAD

**Issue by m0ru** [researchstudio-sat/webofneeds#546]
Here's a few enhancements we could make to drastically increase the page-load duration, roughly sorted by how much effort they are to implement:
1. Minify the bundle (expecting 50%-70% saved in size)
2. Make sure server gzips the bundle
3. Make sure the big bundle is cached by the browser (saves time on reloads)
4. Everything loaded by the load-action as pre-baked script exporting an initialisation object, included directly in index.jsp. This saves us a server round-trip.
5. Find out which scripts are bloating the bundle's size and replace them with slimmer alternatives.
6. divide the bundle into chunks.first load scripts necessary for the actively displayed view, then load rest in a second pass while the user is still on that initial view.

**Comment by fkleedorfer** [comment]
This is really a showstopper for many people.

**Comment by m0ru** [comment]
An first step that we should take before starting with service-workers, would be to generate a hash of every built `bundle.js` and append that to it's name and statically generate `index.html` with `<script src="bundle-a7asd875asdf.js"></script>`. Then we can set the caching-policy with which it's served to `"forever"`.

**Comment by m0ru** [comment]
We could add the [clojure-compiler](https://developers.google.com/closure/compiler/) to the build. Among other things it does limited tree shaking and minification. For a different project I'm currently working on, this reduced the app-bundle size from 2.2MB to 550kB (140kB after gzipping). Dunno if we can't also get babel/jspm to minify/tree-shake.

**Comment by m0ru** [comment]
Also, we should go over our javascript dependencies, check their file-size and try to replace the larger ones with slimmer alternatives.

**Comment by m0ru** [comment]
Related issues:

* Optimize landing page #987
* Optimise images to speed up page-load #978
* Tidy out dependencies in owner-webapp #1047

**Comment by m0ru** [comment]
As for minification: there's a jspm-option for that. Just add `minify: true` to the call to `gulp_jspm(..)` in `gulpfile.js`. This will cause some compilation problems though, that need to be fixed (e.g. missing semicolons).

**Comment by m0ru** [comment]
jspm allows us to split bundle. If angular allows adding components later at run-time, we could pull the landing page it's dependencies into a seperate bundle and load the rest of the app in a second pass. Dunno how we'd react to people entering via the create-need or post-detail pages? have a initial-deferred pair for every possible entry-point?

**Comment by fkleedorfer** [comment]
The worst part is loading all the data. Possible approaches:
1. dropping the in-browser rdf store and making the owner webapp server side fatter/smarter
2. Figuring out what ingormation is needed per need and which needs are needed, and load only that

**Comment by m0ru** [comment]
Some enhancements discussed during today's jour fixe:

* Our fonts (e.g. Open Sans) seem to have a significant page-load performance (~400k). We should try removing font-weights we don't use and if necessary loading those the google-fonts cdn.
* In several of the views (e.g. landingpage, create-need, about) the rdf-store (~300k minified) isn't needed immediatly. We could delay loading it

<wip>

**Comment by m0ru** [comment]
screenshots of the state of things at 69de16c, loading the app with throttling set to "fast 3g" (1Mb/s down, 750kb/s up, 40ms RTT):

**on the landingpage without an account (13s):**

![loading_landingpage_-_good_3g](https://user-images.githubusercontent.com/1730492/30125633-4b3c6742-9339-11e7-9b94-80d9853cf306.PNG)

note that the map loads tiles towards the end, even though it's not visible by default. performance wise this is actually good, should the user continue with a view with a visible map, but has the downside of occuring additional data-transmission (which should be avoided due to causing additional charges)

roughly **13s** till the app has finished loading as far as the user could ever notice, most of which is spent loading data over the network (as there's almost no gaps between the requests)

**on the feed, signed in, with 5 owned needs (18.5s):**

![loading_feed_with_5_posts_-_good_3g](https://user-images.githubusercontent.com/1730492/30125581-24061af6-9339-11e7-860d-59f7361c83bc.PNG)

loading the needs and connections incurs an additional 6.5s (onto a slightly lower base of 12s) for a total of around 18.5s in this case

**Comment by m0ru** [comment]
also rather interesting,

**on the feed, signed in, with 5 owned needs, *repeated visit* (22s):**

![loading_feed_with_5_posts_-_good_3g_-_repeated_visit](https://user-images.githubusercontent.com/1730492/30126842-92b30474-933d-11e7-9245-7e72f06d445b.PNG)

The sprite and the `<script>`-included dependencies are cached now. For some reason however, it redownloads them -- sequentially -- thus causing the even higher load time.

**Comment by m0ru** [comment]
To summarize possible approaches, for general page-load performance we have:

* make sure bundle starts loading asap (e.g. via `<link prefetch...>`)
* remove unused font-weights
* inline styles (?)
* inline fonts (?)
* inline bundle (?)
* make node return need, connections and counterparts a single round-trip (if possible)
* add tree-shaking (e.g. via the clojure compiler or by figuring out how to do this via jspm)
* split bundle, per view into things needed *now* and needed *later*
* look for slimmer alternatives for some of our dependencies (or split off the parts we need from them)
* selective loading of needs for >5 needs and >10 connections
* make sure server uses gzip
* prerender, serve static html and inflate on client (reduces time to meaningful render, not time to interactive; afair not possible in angular)

For repeated visits we could do the following:

* figure out why `<script>`-includes are shown as cached but still redownloaded
* cache needs/connections/events on owner-server and add them to index.hmtl when serving it
* cache needs/connections/events locally in the idb via webworkers (or in the app)
* ensure bundle can be cached / chache it via webworkers

First I'd like to diagnose how much size the dependencies in our package json and our app-folder are contributing to the bundle though.

**Comment by m0ru** [comment]
For when we go super-perfomance <https://hpbn.co/>is a good ressource to read. Most of it's rather neglectable though for us, compared to the bundle-size and the time required to fetch needs and everythin attached to them.

**Comment by m0ru** [comment]
For analyzing our dependencies contribution to bundle-size there's [this nice app](https://www.npmjs.com/package/webpack-bundle-analyzer) (it would require setting up an alternative, webpack-based build though)

**Comment by heikofriedrich** [comment]
hier sind die top punkte komprimierung von js, aufspalten und parallel laden von js und js reduzieren: 
https://developers.google.com/speed/pagespeed/insights/?url=www.matchat.org

**Comment by m0ru** [comment]
One of our larger dependencies and the only dependency that's larger than our app is angular, that we mostly use as a rendering layer. In this capacity we could replace it with just about anything else and save significantly in size (e.g. react, preact, riot, etc). 
 
```
   8276 preact.min.js
  20772 preact.js
  23040 react.min.js
 145032 react.js
 168828 angular.min.js
 380532 app/ (rough minification estimate*)
 962560 app/ (unminified)
1270864 angular.js
```

* using the more realistic factor in [this quick and rough estimation calculation](https://gist.github.com/m0ru/5c7e99621007f355020a41eb1615e943)

**Comment by m0ru** [comment]
@heikofriedrich ah, this answers the question whether or not our server uses gzip -- apparently it doesn't! That should be a rather low hanging fruit indeed we've got there :)

**Comment by heikofriedrich** [comment]
yeah, this google develeoper performance site looks quite useful to me, maybe we could use this to asses how successful our performance optimizations were in the end

**Comment by m0ru** [comment]
in the inlined redux-utils:

export function deletemeeee () {
    console.log('BAAAAAARRRRR');
As for tree-shaking: [this page](https://jspm.io/0.17-beta-guide/static-builds-with-rollup-optimization.html) of the jspm-0.17-beta documentation mentions tree-shaking. I've openened a branch (`feat__treeshaking-with-bower`, named to avoid clashing with a failed branch off of `refac__removed-bower`). In that branch I added jspm-0.17 to the repo and in the powershell (necessary to see the color-highlighting in the output) bundled the app via 

```sh
 jspm build app/app_jspm.js generated/build.js --minify
```

This was done using the CLI, as gulp-jspm doesn't support jspm-0.17 yet and thus our `bundlejs`-task fails.

Tree-shaking works, as a function I added to `redux-utils.js` -- one of the files inlined by rollup (as is indicated by the intendation and color-highlighting) and thus one of the files benefitting from tree-shaking. The function wasn't contained in the output.

However, the resulting `build.js` is bigger than the `app_jspm.bundle.js` build on master. It appears, that we're already using tree-shaking (the same test works as well in `redux-utils.js` on `master`) and for some reason the jspm-0.17-beta output is about 1M bigger, as can be seen here:

```
-rw-r--r-- 1 ksinger 1049089 1475004 Sep 12 14:53 app_jspm.bundle.js
-rw-r--r-- 1 ksinger 1049089 6425584 Sep 12 14:53 app_jspm.bundle.js.map
-rw-r--r-- 1 ksinger 1049089 2512155 Sep 12 17:14 build.js
-rw-r--r-- 1 ksinger 1049089 3161453 Sep 12 17:14 build.js.map
```

**Comment by m0ru** [comment]
For measuring the actual size-impacts of our dependencies (including *their* dependencies), we could simply change `jspm_config.js` to point to an empty module export for each of these in turn, build the bundle and check the size differences. This should even be scriptable without too much effort.

**Comment by m0ru** [comment]
**landinpage, w/o account, after enhancements (4.5s w/o fonts, 5.5s with fonts):**

![loading_landingpage_-_good_3g_-_a-few-enhancments](https://user-images.githubusercontent.com/1730492/30487206-12d9c07e-9a33-11e7-9971-70a664bc6f89.PNG)

With the following enhancements merged:

> 1. {x} #1201(pr): serve the minified css (we already generate it)
> 1. {x} #1203(pr): make sure bundle starts loading asap (e.g. via `<link prefetch...>`, see number 7 [here](https://hackernoon.com/10-things-i-learned-making-the-fastest-site-in-the-world-18a0e1cdf4a7)).  Update: link-preload/prefetch had no effect, however it was possible to bundle mediumjs/rangy/undojs and thus have the bundle load in the first round of ressources after the html is parsed.
> 1. {x} #1198: configure server to use gzip ([expected size-reduction](https://developers.google.com/speed/pagespeed/insights/?url=www.matchat.org) of bundle of 1M/ by 73%)
> 1. {x} #1199(pr): use gzip-compression for fonts 
> 1. {x} #1205(pr): remove unused font-weights
> 1. {x} #1205(pr): non-blocking font-loading / using fall-back fonts
> 1. {x} #1205(pr): woff/woff2 fonts (~10% of the size)
> 1. {x} #1210(pr): gzip svgs
> 1. {x} #1203(pr): inline* or bundle `<script>`-includes

We're now down to **4.5s (w/o fonts)** and **5.5s (with fonts)** on "fast 3G"

**Comment by m0ru** [comment]
As for in-browser-caching: Chrome seems to have troubles caching `won.min.css` and occassionally `OpenSans-*.woff2` and `icon-sprite.svg`. The only difference to e.g. the request for `app_jspm.bundle.js` is that these requests don't contain `cache-control: no-cache` and the responses to them contain `expires: <now+1day>` (while the response with the bundle doesn't).  However I don't see how to add this field in the `<link...>`-tag, i.e. afaict we could only add this by loading the ressources via js or by intercepting the requests and adding the fields.

Other ways to fine-tune the caching:

* set "Vary" to "none" in the responses, as the type of compression/encoding shouldn't change anyway.
* set the `max-age` to one year (controlled via an `ExpiresFilter` in the owner-application's `web.xml`: [see here](https://tomcat.apache.org/tomcat-8.0-doc/config/filter.html#Expires_Filter)) and use revving, i.e. add the a version-number or the build-date to the file-names of the ressources and to the links in `index.html`. A lighter-weight but hackier variant would be something like: `/style.css?modified=20012009`. This only affects re-visits more than a day apart, though.

**Comment by quasarchimaere** [comment]
can we close this item for the time being as we did plenty of performance enhancement as is, and we can open new specific items whenever we feel like it anyway
